File: frontend/src/types/network.ts
export interface NetworkNode {
  id: string;
  label: string;
  entity_type: string;
  risk_score: number;
  risk_level: string;
  alert_count: number;
  transaction_count: number;
  country_code: string | null;
}

export interface NetworkEdge {
  source: string;
  target: string;
  weight: number;
  total_amount: number;
  avg_fraud_score: number;
  is_suspicious: boolean;
}

export interface NetworkGraphStats {
  node_count: number;
  edge_count: number;
  cluster_count: number;
  avg_connections: number;
}

export interface NetworkGraphResponse {
  nodes: NetworkNode[];
  edges: NetworkEdge[];
  stats: NetworkGraphStats;
}

export interface ConnectionDetail {
  connected_entity_id: string;
  name: string;
  type: string;
  transaction_count: number;
  total_amount: number;
  direction: "inbound" | "outbound" | "both";
}

export interface EntityConnectionResponse {
  entity_id: string;
  entity_name: string;
  connections: ConnectionDetail[];
}

File: backend/app/schemas/network.py
from __future__ import annotations

from pydantic import BaseModel, Field


class NetworkNode(BaseModel):
    id: str
    label: str
    entity_type: str
    risk_score: float = 0.0
    risk_level: str = "low"
    alert_count: int = 0
    transaction_count: int = 0
    country_code: str | None = None


class NetworkEdge(BaseModel):
    source: str
    target: str
    weight: int = 1
    total_amount: float = 0.0
    avg_fraud_score: float = 0.0
    is_suspicious: bool = False


class NetworkGraphStats(BaseModel):
    node_count: int = 0
    edge_count: int = 0
    cluster_count: int = 0
    avg_connections: float = 0.0


class NetworkGraphResponse(BaseModel):
    nodes: list[NetworkNode]
    edges: list[NetworkEdge]
    stats: NetworkGraphStats


class ConnectionDetail(BaseModel):
    connected_entity_id: str
    name: str
    type: str
    transaction_count: int
    total_amount: float
    direction: str


class EntityConnectionResponse(BaseModel):
    entity_id: str
    entity_name: str
    connections: list[ConnectionDetail]


class NetworkAnalysisRequest(BaseModel):
    entity_ids: list[str] | None = None
    min_transactions: int = Field(default=2, ge=1)
    include_indirect: bool = False
    max_depth: int = Field(default=2, ge=1, le=5)


class ClusterInfo(BaseModel):
    cluster_id: int
    node_count: int
    risk_score: float
    is_suspicious: bool
    nodes: list[str]


class RiskPath(BaseModel):
    path: list[str]
    risk_score: float


class NetworkAnalysisResponse(BaseModel):
    clusters: list[ClusterInfo]
    risk_paths: list[RiskPath]
    overall_risk: float = 0.0
